### PR TITLE
Credit Scoring AI Act Project

### PR DESCRIPTION
**Pull request by marwan37**
Full credit scoring project showcasing how to build pipelines that align with upcoming EU AI Act regulations. To make the codebase easier to navigate, I added diagrams of all pipelines, the modal deployment app, and the end-to-end flow in the `docs` directory. 

### Features:

- End-to-end pipeline from data loading to model deployment
- EU AI Act compliance artifacts, including Annex IV
- Real-time fairness evaluation
- Human-in-the-loop deployment approvals
- Post-market monitoring with drift detection
- Rich documentation and auto-generated compliance records

### Pipeline runs (with cloud artifact store)
- [Feature engineering run](https://cloud.zenml.io/workspaces/internal/projects/default/runs/3229f39b-db8a-4436-96c3-89effea2b8ad?tab=overview)
- [Training pipeline run](https://cloud.zenml.io/workspaces/internal/projects/6e23c046-6cc4-411c-8b9f-75f0c8a1a818/runs/d34770bf-1bc0-447f-8505-221ed00e08bc?tab=stack)
- [Deployment run](https://cloud.zenml.io/workspaces/internal/projects/default/runs/2491f554-cf31-4085-aab1-cd177baa7ebd?tab=overview)

#### Feature Engineering Pipeline
- **Article 10 (Data Governance)**: Tracked in `data_loader` and `data_preprocessor` (provenance, quality, transformations)
- **Article 12 (Record-keeping)**: Handled by `generate_compliance_metadata` (detailed preprocessing logs)
- **Article 15 (Accuracy)**: Captured in `generate_compliance_metadata` through quality checks
    

#### Training Pipeline
- **Article 9 (Risk Management)**: Risk scoring in `risk_assessment`
- **Article 11 (Technical Documentation)**: Model rationale and design captured during training
- **Article 15 (Accuracy)**: Metrics logged in `evaluate_model`

#### Deployment Pipeline
- **Article 14 (Human Oversight)**: Manual approval in `approve_deployment`
- **Article 17 (Post-market Monitoring)**: Drift detection in `post_market_monitoring`
- **Article 18 (Incident Reporting)**: Tied into `modal_deployment` flow

### CODE REVIEW

File: requirements.txt
Comment (strickvl) [inline requirements.txt:1]: First review comment I noticed is that it is uninstallable for someone using an intel Macbook:

  × No solution found when resolving dependencies:
  ╰─▶ Because tensorflow==2.19.0 has no wheels with a matching platform tag (e.g., `macosx_15_0_x86_64`) and only tensorflow<=2.19.0 is available, we can
      conclude that tensorflow>=2.19.0 cannot be used.
      And because you require tensorflow>=2.19.0, we can conclude that your requirements are unsatisfiable.

      hint: Wheels are available for `tensorflow` (v2.19.0) on the following platforms: `manylinux_2_17_aarch64`, `manylinux_2_17_x86_64`,
      `manylinux2014_aarch64`, `manylinux2014_x86_64`, `macosx_12_0_arm64`, `win_amd64`

Comment (strickvl) [reply requirements.txt:1]: Actually not quite sure where tensorflow is needed in this project.

Comment (marwan37) [reply requirements.txt:1]: I've just removed torch and tensorflow. They were leftovers from the `post_run_annex` step.

Comment (strickvl) [reply requirements.txt:1]: Yes thanks. it works now.

**Review by strickvl — COMMENTED** [review]
Ok so I had a look back at the AI Act's provisions and cross-referenced what's here in the pipelines and I can see some things where we can maybe be more clear for users.

## Core goals

- obviously people will have to do their due diligence, but we can showcase some best practices here to make their lives a bit easier.
- the main thing is that people should take a look at get an idea of the kinds of things they can be doing with ZenML (and know what are the other things that ZenML isn't covering)

## Missing pieces

- Article 8 states you need a quality-management system (QMS) with written procedures + roles listed. We should mention this in the README at least as something important that they have to have.
- `risk_assessment.py` writes a risk register and logs scores, but lacks hazard identification & linkage to mitigations. Not sure if we need the mitigations, but possible hazard identification?
- For technical documentation (article 11), we have `post_run_annex.py` which renders the Jinja template but the generated reports show many 'Required' sections. I'm wondering if there's some way we could either offer some sort of draft version of this with everything completed?
- Art 12 and 19 are all about Automatic logs + record-keeping. Wondering if the README doesn't have to say something about how you set up your artifact store (on some cloud provider) such that you have >= 10 year retention and to make it tamper-proof etc, or at least mention that this is what is expected? (or maybe this is just in the blog?)
- Article 16 has certain obligations around CE declarations etc, so perhaps we could also auto-generate declaration templates for CE? like a `declaration_of_conformity.md` from a template.
- I'm also wondering whether we could turn the `risk_register.xlsx` into a Streamlit dashboard for visibility for non-technical stakeholders?
- I think probably we can commit the rendered Annex IV markdown to `docs/releases/<run_id>` so every run is immutable?
- We could use https://github.com/anchore/syft to generate + log SBOM artifacts that capture container image digests + dependency hashes for Article 15 robustness evidence?
- in general even though there is a good deal of boilerplate to the reports, I think we can actually make them useful as well, by making sure they have good navigation + layout + styling. (also maybe we could have a mini paragraph explaining a very short TLDR for what the section is about + what information is in there or something.)

## Obvious Quick Wins

- whatever reports we output to local filesystem should also be output as HTML artifacts so that they can be viewed in the dashboard
- the reports don't look super nice, so we can make use of markdown tables at the least, and maybe some other things to make it just look much nicer
- probably the reports could use a table of contents at the top
- run all the pipelines using a cloud artifact store so that logs / artifacts are visible in the dashboard
- links to the relevant parts of the AI act in the readme (i.e. to the articles referenced)
- make the project look nice. maybe some screenshots from the soon-to-be beautiful reports that get generated?

## ZenML Internal Knowledge

- it'd be great for our own internal documentation (perhaps as something growth team could reference / use, when needed) to have a nice table showing exactly the specific obligations under the AI Act in terms of what you have to produce / generate, and then another column showing how you can do this with ZenML (bonus: perhaps this points to specific implementation details in the files once the project is complete)

File: README.md [hunk 1]
```diff
@@ -1,88 +1,185 @@
-# Credit‑Scoring AI‑Act Demo
+# Credit‑Scoring EU AI Act Demo
 
-> Demo project to showcase EU AI Act compliance artifacts with ZenML.
+> A ZenML‑powered end‑to‑end credit‑scoring workflow that automatically generates the technical evidence required by the EU AI Act.
```
Comment (strickvl) [inline README.md]: link to the act

Comment (strickvl) [reply README.md]: also maybe link to our blog and add a one sentence something about why we care about it? (Lots of people haven't heard of the AI Act or know how it affects them etc, so we should make that clear, right?)

File: README.md [hunk 2]
```diff
@@ -1,88 +1,185 @@
...
-# -> approval gate records reviewer
-# -> drift monitor cronjob scheduled
-# -> compliance/records/ contains run artifacts
-# -> compliance/manual_fills/ contains required inputs
-```
+Pipeline configurations are stored in the `src/configs/` directory:
+
+- `feature_engineering.yaml`
+- `training.yaml`
+- `deployment.yaml`
```
Comment (strickvl) [inline README.md]: maybe link to the files themselves?

File: README.md [hunk 3]
```diff
@@ -1,88 +1,185 @@
...
+
+![Modal Deployment](assets/modal-deployment.png)
+
+The project implements a serverless deployment using Modal with comprehensive monitoring and incident reporting capabilities:
+
+- FastAPI application with documented endpoints
+- Automated model and preprocessing pipeline loading
+- Drift detection and incident reporting
+- Standardized storage paths for compliance artifacts
+- Complete incident reporting system for Article 18 compliance
```
Comment (strickvl) [inline README.md]: I have a feeling that maybe the claims in here need toning down...  like what drift detection and incident reporting are we doing here in the deployment pipeline (I couldn't find that in code...)

File: README.md
Comment (strickvl) [inline README.md:1]: could you add links to some representive runs from wherever you ran them (on cloud-artifact-store-backed stacks)?

Comment (strickvl) [reply README.md:1]: (add links in the PR description, I mean...)

File: README.md [hunk 4]
```diff
@@ -1,88 +1,185 @@
...
+
+## 🔗 EU AI Act Compliance Mapping
+
+For a complete overview of the EU AI Act compliance mapping, refer to the [COMPLIANCE.md](COMPLIANCE.md) file.
```
Comment (strickvl) [inline README.md]: 'an overview of the mapping' you mean 'an overview of the ways this pipeline maps to the articles of the EU AI Act' or something? maybe update to say that.

File: README.md [hunk 5]
```diff
@@ -1,88 +1,185 @@
...
+
+The **releases/** directory contains automatically generated artifacts for each pipeline run, including:
+
+- Annex IV technical documentation (annex_iv_cs_deployment.md)
+- Software Bill of Materials (sbom.json)
+- Model performance metrics (evaluation_results.yaml)
+- Risk assessment scores (risk_scores.yaml)
+- Data profiling report (whylogs_profile.html)
+- Monitoring configuration (monitoring_plan.json)
+- Git repository information (git_info.md)
```
Comment (strickvl) [inline README.md]: let's link to these files instead of just showing the file name?

File: README.md
Comment (strickvl) [inline README.md:1]: also assuming you made the reports look much nicer, could we include some screenshots of the nice rendered reports in this README as well?

File: assets/feature-engineering-pipeline.png
Comment (strickvl) [inline assets/feature-engineering-pipeline.png:1]: I think probably all these images should go through tinypng.com and be resized down?

File: docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/annex_iv_cs_deployment.md [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+
+- **cs_deployment-2025_05_18-15_59_04_235495** (`e3056fd1`) - 2025-05-18 15:59:05 *[completed]*
+
+- **cs_deployment-2025_05_18-15_58_11_396593** (`790ce633`) - 2025-05-18 15:58:12 *[failed]*
+
+- **cs_deployment-2025_05_18-15_45_12_611631** (`955e15c3`) - 2025-05-18 15:45:13 *[completed]*
+
+- **cs_deployment-2025_05_18-15_29_30_867372** (`a109c9ec`) - 2025-05-18 15:29:31 *[completed]*
+
+- **cs_deployment-2025_05_18-15_28_29_739734** (`131d10f4`) - 2025-05-18 15:28:30 *[failed]*
```
Comment (strickvl) [inline docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/annex_iv_cs_deployment.md]: this seems ripe for a table. did you try that?

File: docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/annex_iv_cs_deployment.md [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+
+**Artifact_store:**
+
+- **default** (local, built-in)
+
+
+
+**Orchestrator:**
+
+- **default** (local, built-in)
```
Comment (strickvl) [inline docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/annex_iv_cs_deployment.md]: I think we want to also include the component ids. And for each component we can include the kind of information you see when you run `zenml orchestrator describe`, right?

File: docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/annex_iv_cs_deployment.md [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+| xlsxwriter | >=3.2.3 |
+| zenml | >=0.82.1 |
+
+
+### 1(d) Deployment Forms
+
+**Deployment Configuration:**
+- **Type:** Modal + FastAPI (Serverless API deployment)
+- **Environment:** Production
+- **Scaling:** Automatic
```
Comment (strickvl) [inline docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/annex_iv_cs_deployment.md]: probably needs ids here of the modal deployment etc>?

File: docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/git_info.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+**Commit Date:** 2025-05-18T21:45:40
+
+**Author:** marwan37 <marwan.ext@zenml.io>
+
+**Message:**
+```
+update annex iv template and sample inputs
+
+```
```
Comment (strickvl) [inline docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/git_info.md]: what's this?

File: src/configs/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+# Pipeline options
+enable_cache: False
+enable_artifact_metadata: True
+
+steps:
+  modal_deployment:
+    settings:
+      resources:
+        cpu_count: 6
+        memory: "14GB"
```
Comment (strickvl) [inline src/configs/deployment.yaml]: why do we need so much?

File: src/configs/feature_engineering.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
+# environment configuration
+settings:
+  docker:
+    required_integrations:
+      - sklearn
+      - pandas
+    requirements:
+      - pyarrow
```
Comment (strickvl) [inline src/configs/feature_engineering.yaml]: how come can't just specify in the requirements.txt?

File: src/configs/training.yaml [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    - "NAME_EDUCATION_TYPE"
+    - "NAME_FAMILY_STATUS"
+    - "NAME_HOUSING_TYPE"
+
+steps:
+  train_model:
+    settings:
+      resources:
+        cpu_count: 6
+        memory: "14GB"
```
Comment (strickvl) [inline src/configs/training.yaml]: same here

File: src/constants.py
Comment (strickvl) [inline src/constants.py:1]: haha wow. so at least some of this probably belongs in the config files instead of constants.py. WDYT?

File: src/utils/model_definition.py
Comment (strickvl) [inline src/utils/model_definition.py:1]: somehow feel like this should belong closer to the place where it's referenced. feels not really like a 'util'

File: src/utils/template.py
Comment (strickvl) [inline src/utils/template.py:1]: maybe improve the type hints to be more specific here?

File: streamlit_app.py
Comment (strickvl) [inline streamlit_app.py:1]: maybe screenshot of the this dashboard view somewhere in the README if it looks nice?

File: app/modal_deployment.py [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+        modal.Image.debian_slim(python_version=python_version)
+        .pip_install(
+            "scikit-learn",
+            "joblib",
+            "pandas",
+            "numpy",
+            "requests",
+            "whylogs",
+            "fastapi[standard]",
+            "uvicorn",
```
Comment (strickvl) [inline app/modal_deployment.py]: why aren't we passing these in from the requirements file etc?

File: README.md
Comment (strickvl) [inline README.md:1]: going through the code it seems that whylogs is being used. should probably mention that somewhere in the README? also maybe needs a 'how to run this' section in the README which includes installing whylogs etc?

**Review by strickvl — CHANGES_REQUESTED** [review]

**Comment by strickvl** [comment]
Also not sure whether this should be in its own repo vs being a new project...

File: README.md (hunk 3 above)
Comment (marwan37) [reply README.md]: there is some very basic incident reporting implemented, but you're right about data drift. I'll scratch that part off and tone down the rest.

File: docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/annex_iv_cs_deployment.md (hunk 6 above)
Comment (marwan37) [reply docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/annex_iv_cs_deployment.md]: no, because I didn't even realize I had that many already 😅. Good idea though.

File: docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/git_info.md (hunk 9 above)
Comment (marwan37) [reply docs/releases/e7243682-a6f7-4f1d-b8ff-dc8da754994a/git_info.md]: the last git commit message before deployment.

File: src/configs/deployment.yaml (hunk 10 above)
Comment (marwan37) [reply src/configs/deployment.yaml]: these are just sample inputs, but I can make them more realistic

File: src/constants.py
Comment (marwan37) [reply src/constants.py:1]: possibly, yes :)

File: README.md
Comment (marwan37) [reply README.md:1]: whylogs is in the requirements, no need to install it separately. But i can add a little block in the README about it

File: app/modal_deployment.py (hunk 13 above)
Comment (marwan37) [reply app/modal_deployment.py]: to avoid including zenml

**Comment by marwan37** [comment]
@strickvl I've implemented 95% of your suggestions 😅, and also may have gone overboard with the streamlit app but I feel like it's much more useful + pretty cool now. I'll transfer this repo + commit history to zenml_projects and create a PR there.

Closing this PR --> Moved here: https://github.com/zenml-io/zenml-projects/pull/223
